### PR TITLE
🚀 Implement Persistent Settings Storage with Firebase Integration (#184)

### PR DESCRIPTION
**Pull request by mehdixboutaleb**
**### 🚀 Implement Persistent Settings Storage with Firebase Integration (#184)**

This pull request addresses issue #184 by adding methods to store user settings parameters on Firebase, ensuring that user preferences are saved and retrieved effectively.

📌 Summary of Changes

🏠 MainActivity Updates

Settings Screen Support: Updated MainActivity to accommodate the new structure of the settings screen, integrating it seamlessly into the app flow.
ViewModel and Repository Integration: Injected SettingsViewModel and SettingsRepository into MainActivity to manage state and data operations for the settings screen.

🗄️ SettingsRepository Implementation

Firebase Firestore Integration: Created SettingsRepository to handle all interactions with Firebase Firestore for storing and retrieving user settings.
Methods Implemented:
setPrivacySettings(userId: String, value: Boolean)
getPrivacySettings(userId: String): Boolean
setSelectedLanguage(userId: String, language: String)
getSelectedLanguage(userId: String): String
setSelectedTheme(userId: String, theme: String)
getSelectedTheme(userId: String): String
Exception Handling: Wrapped Firebase operations in try-catch blocks to handle any potential exceptions, ensuring robustness.

🧠 SettingsViewModel Enhancements

State Management with StateFlow: Utilized MutableStateFlow to manage privacySettings, selectedLanguage, and selectedTheme, providing reactive updates to the UI.
Data Synchronization: Ensured that changes in settings are immediately reflected in Firebase and the app's UI.
Authentication Checks: Added checks for user authentication before performing any data operations, logging appropriate messages when the user is not authenticated.

🎛️ Settings Screen Modifications

Storage Feature Integration: Modified SettingsScreen to implement the storage feature, enabling users to save their preferences persistently.
User Experience Improvements: Enhanced the UI responsiveness and ensured that settings are accurately displayed and updated.

🧪 Testing and Validation

🔍 SettingsRepositoryTest

Comprehensive Test Cases: Wrote tests for all methods in SettingsRepository to verify correct data storage and retrieval from Firebase.
Mocking Firebase Interactions: Used MockK to mock FirebaseFirestore, DocumentReference, and DocumentSnapshot, ensuring tests do not interact with the real database.
Edge Case Testing: Included tests for scenarios where data might be null or exceptions are thrown, validating default values and exception handling.

🔍 SettingsViewModelTest
Initialization Tests: Confirmed that the ViewModel correctly loads user settings upon initialization.
Update Methods Testing: Verified that updating settings through the ViewModel updates both the local state and Firebase.
Exception Handling Tests: Tested the ViewModel's behavior when exceptions occur during data loading or saving.
Unauthenticated User Handling: Ensured that the ViewModel behaves correctly when the user is not authenticated, avoiding unnecessary Firebase calls.

📂 Files Modified

MainActivity.kt: Updated to include the new settings screen and associated ViewModel and Repository.
SettingsRepository.kt: Implemented methods for interacting with Firebase Firestore to store and retrieve settings.
SettingsRepositoryTest.kt: Added tests for SettingsRepository methods, covering various scenarios and exceptions.
SettingsViewModel.kt: Enhanced to manage settings state and interact with the repository.
SettingsViewModelTest.kt: Expanded tests to cover new ViewModel functionalities and edge cases.
SettingsScreen.kt: Modified to integrate the storage feature, reflecting real-time updates from the ViewModel.

📈 Impact

Persistent User Preferences: Users' settings are now stored in Firebase, allowing for consistent experiences across sessions and devices.
Improved Reliability: Robust exception handling and comprehensive testing increase the reliability of settings management.
Scalability: The implementation lays the groundwork for future settings and preferences to be added with ease.

### CODE REVIEW

File: app/src/test/java/com/github/se/eduverse/viewmodel/SettingsViewModelTest.kt [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    // Verify StateFlow retains default values
+    assertEquals(true, newViewModel.privacySettings.value)
+    assertEquals("English", newViewModel.selectedLanguage.value)
+    assertEquals("Light", newViewModel.selectedTheme.value)
+  }
+
+  @Test
+  fun updatePrivacySettings_handlesException() = runTest {
```
Comment (AKallala) [inline app/src/test/java/com/github/se/eduverse/viewmodel/SettingsViewModelTest.kt:157]: You can add the same test for other functions to increase coverage

File: app/src/test/java/com/github/se/eduverse/repository/SettingsRepositoryTest.kt [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+
+    // Act
+    repository.setSelectedTheme(userId, newTheme)
+
+    // Assert
+    coVerify(exactly = 1) { documentReference.set(data, SetOptions.merge()).await() }
+  }
+
+  @Test
+  fun `getPrivacySettings handles exceptions`() = runTest {
```
Comment (AKallala) [inline app/src/test/java/com/github/se/eduverse/repository/SettingsRepositoryTest.kt:227]: You should add a similar test for other functions to increase coverage

**Review by AKallala — CHANGES_REQUESTED** [review]
I suggested some changes, otherwise your code is of good quality, well done ! 
Also try to increase the code coverage in the tests.

File: app/src/main/java/com/github/se/eduverse/ui/setting/Setting.kt [hunk 3]
```diff
@@ -105,7 +107,7 @@ fun SettingsScreen(navigationActions: NavigationActions) {
               label = "Language",
               selectedOption = selectedLanguage,
               options = listOf("Français", "English"),
```
[the remark this answers is not in the document]
Comment (mehdixboutaleb) [reply app/src/main/java/com/github/se/eduverse/ui/setting/Setting.kt:109]: I think it's better to leave it like this, in order to not forget that it could be a possible future implementation. If we decide to not implement that I will make sure to delete this part.

**Review by AKallala — APPROVED** [review]
LGTM
